src/zarr/core/group.py

```diff
@@ -2765,6 +2836,327 @@ def array(
+
+                attrs_done = _join_paths([node_name, ZATTRS_JSON]) in created_object_keys
+
+                if isinstance(meta_out, GroupMetadata):
+                    meta_done = _join_paths([node_name, ZGROUP_JSON]) in created_object_keys
+                else:
+                    meta_done = _join_paths([node_name, ZARRAY_JSON]) in created_object_keys
+
+                if meta_done and attrs_done:
+                    node_store_path = store_path / node_name
+                    if isinstance(meta_out, GroupMetadata):
+                        yield AsyncGroup(metadata=meta_out, store_path=node_store_path)
+                    else:
+                        yield AsyncArray(metadata=meta_out, store_path=node_store_path)
+                continue
+
+
+T = TypeVar("T")
+
+
+def _get_roots(
+    data: dict[str, GroupMetadata | ArrayV2Metadata | ArrayV3Metadata],
+) -> tuple[str, ...]:
+    """
+    Return the keys of the root(s) of the hierarchy. A root is a key with the fewest number of
+    path segments.
+    """
+    if "" in data:
+        return ("",)
+    keys_split = sorted((key.split("/") for key in data), key=len)
+    groups: defaultdict[int, list[str]] = defaultdict(list)
+    for key_split in keys_split:
+        groups[len(key_split)].append("/".join(key_split))
+    return tuple(groups[min(groups.keys())])
+
+
+def _join_paths(paths: Iterable[str]) -> str:
+    """
+    Filter out instances of '' and join the remaining strings with '/'.
+
+    Because the root node of a zarr hierarchy is represented by an empty string,
+    """
+    return "/".join(filter(lambda v: v != "", paths))
+
+
+def _parse_hierarchy_dict(
+    data: Mapping[str, GroupMetadata | ArrayV2Metadata | ArrayV3Metadata],
+) -> dict[str, GroupMetadata | ArrayV2Metadata | ArrayV3Metadata]:
+    """
+    If the input represents a complete Zarr hierarchy, i.e. one with no implicit groups,
+    then return an identical copy of that dict. Otherwise, return a version of the input dict
+    with groups added where they are needed to make the hierarchy explicit.
+
+    For example, an input of {'a/b/c': ArrayMetadata} will result in a return value of
+    {'a': GroupMetadata, 'a/b': GroupMetadata, 'a/b/c': ArrayMetadata}.
+
+    The input is also checked for the following conditions, and an error is raised if any
+    of them are violated:
+
+    - No arrays can contain group or arrays (i.e., all arrays must be leaf nodes).
+    - All arrays and groups must have the same ``zarr_format`` value.
+
+    This function ensures that the input is transformed into a specification of a complete and valid
+    Zarr hierarchy.
+    """
+
+    observed_zarr_formats: dict[ZarrFormat, list[str | None]] = {2: [], 3: []}
+
+    # We will iterate over the dict again, but a full pass here ensures that the error message
+    # is comprehensive, and I think the performance cost will be negligible.
+    for k, v in data.items():
+        observed_zarr_formats[v.zarr_format].append(k)
```

Comment:
can we short-circuit inside the loop and raise on the first instance of `zarr_format` that is not equal to `next(data.items())[1].zarr_format`?

Reply:
we definitely could, but we get a much nicer error message if we can traverse the entire proposed hierarchy and identify all the problematic nodes.